File: sdk/python/src/aion_sdk/qa.py
"""QA API interface."""

from typing import List, Optional, Dict, Any
from uuid import UUID

from aion_sdk.models import (
    QASession,
    QARequest,
    QATestType,
    QAConfiguration,
    QAResults,
    CoverageReport,
    QualityMetrics,
    PaginatedResponse,
)


class QAAPI:
    """QA API interface for automated testing and quality assurance."""

    def __init__(self, client: "AionClient") -> None:
        """Initialize the QA API."""
        self._client = client

    async def start_session(self, request: QARequest) -> QASession:
        """Start a new QA session.

        Args:
            request: QA session request

        Returns:
            The created QA session
        """
        data = await self._client.post("/api/v1/qa/sessions", json_data=request.model_dump())
        return QASession.model_validate(data)

    async def get_session(self, session_id: UUID) -> QASession:
        """Get QA session details.

        Args:
            session_id: The session ID

        Returns:
            The QA session details
        """
        data = await self._client.get(f"/api/v1/qa/sessions/{session_id}")
        return QASession.model_validate(data)

    async def list_sessions(
        self,
        project_id: UUID,
        page: int = 1,
        per_page: int = 20,
    ) -> PaginatedResponse[QASession]:
        """List QA sessions for a project.

        Args:
            project_id: The project ID
            page: Page number (default: 1)
            per_page: Items per page (default: 20)

        Returns:
            Paginated list of QA sessions
        """
        params = {
            "project_id": str(project_id),
            "page": page,
            "per_page": per_page,
        }

        data = await self._client.get("/api/v1/qa/sessions", params=params)
        return PaginatedResponse.model_validate(data)

    async def stop_session(self, session_id: UUID) -> None:
        """Stop a running QA session.

        Args:
            session_id: The session ID
        """
        await self._client.post(f"/api/v1/qa/sessions/{session_id}/stop", json_data={})

    async def cancel_session(self, session_id: UUID) -> None:
        """Cancel a QA session.

        Args:
            session_id: The session ID
        """
        await self._client.post(f"/api/v1/qa/sessions/{session_id}/cancel", json_data={})

    async def get_results(self, session_id: UUID) -> QAResults:
        """Get session results.

        Args:
            session_id: The session ID

        Returns:
            The QA results
        """
        data = await self._client.get(f"/api/v1/qa/sessions/{session_id}/results")
        return QAResults.model_validate(data)

    async def get_logs(
        self,
        session_id: UUID,
        level: Optional[str] = None,
        page: int = 1,
        per_page: int = 50,
    ) -> PaginatedResponse[Dict[str, Any]]:
        """Get session logs.

        Args:
            session_id: The session ID
            level: Log level filter ('debug', 'info', 'warn', 'error')
            page: Page number (default: 1)
            per_page: Items per page (default: 50)

        Returns:
            Paginated list of log entries
        """
        params = {
            "page": page,
            "per_page": per_page,
        }
        if level:
            params["level"] = level

        data = await self._client.get(f"/api/v1/qa/sessions/{session_id}/logs", params=params)
        return PaginatedResponse.model_validate(data)

    async def run_unit_tests(
        self,
        project_id: UUID,
        config: Optional[QAConfiguration] = None,
    ) -> QASession:
        """Run unit tests for a project.

        Args:
            project_id: The project ID
            config: Optional QA configuration

        Returns:
            The QA session
        """
        request = QARequest(
            project_id=project_id,
            test_type=QATestType.UNIT,
            configuration=config,
        )
        return await self.start_session(request)

    async def run_integration_tests(
        self,
        project_id: UUID,
        config: Optional[QAConfiguration] = None,
    ) -> QASession:
        """Run integration tests for a project.

        Args:
            project_id: The project ID
            config: Optional QA configuration

        Returns:
            The QA session
        """
        request = QARequest(
            project_id=project_id,
            test_type=QATestType.INTEGRATION,
            configuration=config,
        )
        return await self.start_session(request)

    async def run_e2e_tests(
        self,
        project_id: UUID,
        config: Optional[QAConfiguration] = None,
    ) -> QASession:
        """Run end-to-end tests for a project.

        Args:
            project_id: The project ID
            config: Optional QA configuration

        Returns:
            The QA session
        """
        request = QARequest(
            project_id=project_id,
            test_type=QATestType.E2E,
            configuration=config,
        )
        return await self.start_session(request)

    async def run_performance_tests(
        self,
        project_id: UUID,
        config: Optional[QAConfiguration] = None,
    ) -> QASession:
        """Run performance tests for a project.

        Args:
            project_id: The project ID
            config: Optional QA configuration

        Returns:
            The QA session
        """
        request = QARequest(
            project_id=project_id,
            test_type=QATestType.PERFORMANCE,
            configuration=config,
        )
        return await self.start_session(request)

    async def run_security_tests(
        self,
        project_id: UUID,
        config: Optional[QAConfiguration] = None,
    ) -> QASession:
        """Run security tests for a project.

        Args:
            project_id: The project ID
            config: Optional QA configuration

        Returns:
            The QA session
        """
        request = QARequest(
            project_id=project_id,
            test_type=QATestType.SECURITY,
            configuration=config,
        )
        return await self.start_session(request)

    async def run_accessibility_tests(
        self,
        project_id: UUID,
        config: Optional[QAConfiguration] = None,
    ) -> QASession:
        """Run accessibility tests for a project.

        Args:
            project_id: The project ID
            config: Optional QA configuration

        Returns:
            The QA session
        """
        request = QARequest(
            project_id=project_id,
            test_type=QATestType.ACCESSIBILITY,
            configuration=config,
        )
        return await self.start_session(request)

    async def run_comprehensive_tests(
        self,
        project_id: UUID,
        config: Optional[QAConfiguration] = None,
    ) -> QASession:
        """Run comprehensive test suite for a project.

        Args:
            project_id: The project ID
            config: Optional QA configuration

        Returns:
            The QA session
        """
        request = QARequest(
            project_id=project_id,
            test_type=QATestType.COMPREHENSIVE,
            configuration=config,
        )
        return await self.start_session(request)

    async def get_coverage(self, project_id: UUID) -> CoverageReport:
        """Get test coverage report for a project.

        Args:
            project_id: The project ID

        Returns:
            The coverage report
        """
        data = await self._client.get(f"/api/v1/qa/projects/{project_id}/coverage")
        return CoverageReport.model_validate(data)

    async def get_quality_metrics(self, project_id: UUID) -> QualityMetrics:
        """Get quality metrics for a project.

        Args:
            project_id: The project ID

        Returns:
            The quality metrics
        """
        data = await self._client.get(f"/api/v1/qa/projects/{project_id}/quality")
        return QualityMetrics.model_validate(data)

    async def get_recommendations(self, project_id: UUID) -> List[Dict[str, Any]]:
        """Get test recommendations for a project.

        Args:
            project_id: The project ID

        Returns:
            List of recommendations
        """
        return await self._client.get(f"/api/v1/qa/projects/{project_id}/recommendations")

    async def generate_tests(self, request: Dict[str, Any]) -> Dict[str, Any]:
        """Generate test cases for a project.

        Args:
            request: Test generation request

        Returns:
            Test generation response
        """
        return await self._client.post("/api/v1/qa/generate-tests", json_data=request)

    async def validate_config(self, config: QAConfiguration) -> Dict[str, Any]:
        """Validate test configuration.

        Args:
            config: QA configuration to validate

        Returns:
            Validation result
        """
        return await self._client.post("/api/v1/qa/validate-config", json_data=config.model_dump())

    async def get_stats(self, project_id: Optional[UUID] = None) -> Dict[str, Any]:
        """Get QA statistics.

        Args:
            project_id: Optional project ID to filter stats

        Returns:
            QA statistics
        """
        params = {}
        if project_id:
            params["project_id"] = str(project_id)

        return await self._client.get("/api/v1/qa/stats", params=params)

    async def export_results(self, session_id: UUID, format: str = "json") -> bytes:
        """Export test results.

        Args:
            session_id: The session ID
            format: Export format ('json', 'xml', 'html', 'pdf')

        Returns:
            Exported results as bytes
        """
        params = {"format": format}
        response = await self._client._client.get(f"/api/v1/qa/sessions/{session_id}/export", params=params)
        response.raise_for_status()
        return response.content

File: crates/aion-api-client/src/qa.rs
use crate::{types::*, error::Result, client::AionClient};
use uuid::Uuid;

#[derive(Debug, Clone)]
pub struct QaApi {
    client: AionClient,
}

impl QaApi {
    pub(crate) fn new(client: AionClient) -> Self {
        Self { client }
    }

    /// Start a new QA session
    pub async fn start_session(&self, request: QARequest) -> Result<QASession> {
        self.client.post("/api/v1/qa/sessions", &request).await
    }

    /// Get QA session details
    pub async fn get_session(&self, session_id: Uuid) -> Result<QASession> {
        let path = format!("/api/v1/qa/sessions/{}", session_id);
        self.client.get(&path).await
    }

    /// List QA sessions for a project
    pub async fn list_sessions(&self, project_id: Uuid, params: Option<PaginationParams>) -> Result<PaginatedResponse<QASession>> {
        let mut path = format!("/api/v1/qa/sessions?project_id={}", project_id);

        if let Some(params) = params {
            if let Some(page) = params.page {
                path.push_str(&format!("&page={}", page));
            }
            if let Some(per_page) = params.per_page {
                path.push_str(&format!("&per_page={}", per_page));
            }
        }

        self.client.get(&path).await
    }

    /// Stop a running QA session
    pub async fn stop_session(&self, session_id: Uuid) -> Result<()> {
        let path = format!("/api/v1/qa/sessions/{}/stop", session_id);
        self.client.post::<serde_json::Value, _>(&path, &serde_json::json!({})).await?;
        Ok(())
    }

    /// Cancel a QA session
    pub async fn cancel_session(&self, session_id: Uuid) -> Result<()> {
        let path = format!("/api/v1/qa/sessions/{}/cancel", session_id);
        self.client.post::<serde_json::Value, _>(&path, &serde_json::json!({})).await?;
        Ok(())
    }

    /// Get session results
    pub async fn get_results(&self, session_id: Uuid) -> Result<QAResults> {
        let path = format!("/api/v1/qa/sessions/{}/results", session_id);
        self.client.get(&path).await
    }

    /// Get session logs
    pub async fn get_logs(&self, session_id: Uuid, params: Option<LogParams>) -> Result<PaginatedResponse<QALogEntry>> {
        let mut path = format!("/api/v1/qa/sessions/{}/logs", session_id);
        let mut query_params = Vec::new();

        if let Some(params) = params {
            if let Some(level) = params.level {
                let level_str = match level {
                    LogLevel::Debug => "debug",
                    LogLevel::Info => "info",
                    LogLevel::Warn => "warn",
                    LogLevel::Error => "error",
                };
                query_params.push(format!("level={}", level_str));
            }

            if let Some(page) = params.pagination.page {
                query_params.push(format!("page={}", page));
            }

            if let Some(per_page) = params.pagination.per_page {
                query_params.push(format!("per_page={}", per_page));
            }
        }

        if !query_params.is_empty() {
            path.push('?');
            path.push_str(&query_params.join("&"));
        }

        self.client.get(&path).await
    }

    /// Run unit tests
    pub async fn run_unit_tests(&self, project_id: Uuid, config: Option<QAConfiguration>) -> Result<QASession> {
        let request = QARequest {
            project_id,
            test_type: QATestType::Unit,
            configuration: config,
        };
        self.start_session(request).await
    }

    /// Run integration tests
    pub async fn run_integration_tests(&self, project_id: Uuid, config: Option<QAConfiguration>) -> Result<QASession> {
        let request = QARequest {
            project_id,
            test_type: QATestType::Integration,
            configuration: config,
        };
        self.start_session(request).await
    }

    /// Run end-to-end tests
    pub async fn run_e2e_tests(&self, project_id: Uuid, config: Option<QAConfiguration>) -> Result<QASession> {
        let request = QARequest {
            project_id,
            test_type: QATestType::E2E,
            configuration: config,
        };
        self.start_session(request).await
    }

    /// Run performance tests
    pub async fn run_performance_tests(&self, project_id: Uuid, config: Option<QAConfiguration>) -> Result<QASession> {
        let request = QARequest {
            project_id,
            test_type: QATestType::Performance,
            configuration: config,
        };
        self.start_session(request).await
    }

    /// Run security tests
    pub async fn run_security_tests(&self, project_id: Uuid, config: Option<QAConfiguration>) -> Result<QASession> {
        let request = QARequest {
            project_id,
            test_type: QATestType::Security,
            configuration: config,
        };
        self.start_session(request).await
    }

    /// Run accessibility tests
    pub async fn run_accessibility_tests(&self, project_id: Uuid, config: Option<QAConfiguration>) -> Result<QASession> {
        let request = QARequest {
            project_id,
            test_type: QATestType::Accessibility,
            configuration: config,
        };
        self.start_session(request).await
    }

    /// Run comprehensive test suite
    pub async fn run_comprehensive_tests(&self, project_id: Uuid, config: Option<QAConfiguration>) -> Result<QASession> {
        let request = QARequest {
            project_id,
            test_type: QATestType::Comprehensive,
            configuration: config,
        };
        self.start_session(request).await
    }

    /// Get test coverage report
    pub async fn get_coverage(&self, project_id: Uuid) -> Result<CoverageReport> {
        let path = format!("/api/v1/qa/projects/{}/coverage", project_id);
        self.client.get(&path).await
    }

    /// Get quality metrics
    pub async fn get_quality_metrics(&self, project_id: Uuid) -> Result<QualityMetrics> {
        let path = format!("/api/v1/qa/projects/{}/quality", project_id);
        self.client.get(&path).await
    }

    /// Get test recommendations
    pub async fn get_recommendations(&self, project_id: Uuid) -> Result<Vec<TestRecommendation>> {
        let path = format!("/api/v1/qa/projects/{}/recommendations", project_id);
        self.client.get(&path).await
    }

    /// Generate test cases
    pub async fn generate_tests(&self, request: TestGenerationRequest) -> Result<TestGenerationResponse> {
        self.client.post("/api/v1/qa/generate-tests", &request).await
    }

    /// Validate test configuration
    pub async fn validate_config(&self, config: QAConfiguration) -> Result<ConfigValidationResult> {
        self.client.post("/api/v1/qa/validate-config", &config).await
    }

    /// Get QA statistics
    pub async fn get_stats(&self, project_id: Option<Uuid>) -> Result<QAStats> {
        let path = if let Some(project_id) = project_id {
            format!("/api/v1/qa/stats?project_id={}", project_id)
        } else {
            "/api/v1/qa/stats".to_string()
        };
        self.client.get(&path).await
    }

    /// Export test results
    pub async fn export_results(&self, session_id: Uuid, format: ExportFormat) -> Result<bytes::Bytes> {
        let format_str = match format {
            ExportFormat::Json => "json",
            ExportFormat::Xml => "xml",
            ExportFormat::Html => "html",
            ExportFormat::Pdf => "pdf",
        };

        let path = format!("/api/v1/qa/sessions/{}/export?format={}", session_id, format_str);
        let url = self.client.base_url().join(&path)?;
        let response = self.client.http_client().get(url).send().await?;

        if response.status().is_success() {
            Ok(response.bytes().await?)
        } else {
            Err(crate::error::AionError::Api {
                status: response.status().as_u16(),
                message: "Failed to export results".to_string(),
            })
        }
    }
}

#[derive(Debug, Clone)]
pub struct LogParams {
    pub level: Option<LogLevel>,
    pub pagination: PaginationParams,
}

#[derive(Debug, Clone)]
pub enum LogLevel {
    Debug,
    Info,
    Warn,
    Error,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct QALogEntry {
    pub timestamp: chrono::DateTime<chrono::Utc>,
    pub level: String,
    pub message: String,
    pub context: Option<serde_json::Value>,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct CoverageReport {
    pub overall_coverage: f32,
    pub line_coverage: f32,
    pub branch_coverage: f32,
    pub function_coverage: f32,
    pub files: Vec<FileCoverage>,
    pub uncovered_lines: Vec<UncoveredLine>,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct FileCoverage {
    pub file_path: String,
    pub line_coverage: f32,
    pub branch_coverage: f32,
    pub function_coverage: f32,
    pub total_lines: u32,
    pub covered_lines: u32,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct UncoveredLine {
    pub file_path: String,
    pub line_number: u32,
    pub line_content: String,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct QualityMetrics {
    pub maintainability_index: f32,
    pub cyclomatic_complexity: f32,
    pub code_duplication: f32,
    pub technical_debt_ratio: f32,
    pub test_coverage: f32,
    pub security_rating: SecurityRating,
    pub reliability_rating: ReliabilityRating,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub enum SecurityRating {
    #[serde(rename = "A")]
    A,
    #[serde(rename = "B")]
    B,
    #[serde(rename = "C")]
    C,
    #[serde(rename = "D")]
    D,
    #[serde(rename = "E")]
    E,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub enum ReliabilityRating {
    #[serde(rename = "A")]
    A,
    #[serde(rename = "B")]
    B,
    #[serde(rename = "C")]
    C,
    #[serde(rename = "D")]
    D,
    #[serde(rename = "E")]
    E,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct TestRecommendation {
    pub category: RecommendationCategory,
    pub priority: RecommendationPriority,
    pub title: String,
    pub description: String,
    pub suggested_action: String,
    pub estimated_effort: String,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub enum RecommendationCategory {
    #[serde(rename = "coverage")]
    Coverage,
    #[serde(rename = "performance")]
    Performance,
    #[serde(rename = "security")]
    Security,
    #[serde(rename = "maintainability")]
    Maintainability,
    #[serde(rename = "best_practices")]
    BestPractices,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub enum RecommendationPriority {
    #[serde(rename = "critical")]
    Critical,
    #[serde(rename = "high")]
    High,
    #[serde(rename = "medium")]
    Medium,
    #[serde(rename = "low")]
    Low,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct TestGenerationRequest {
    pub project_id: Uuid,
    pub file_patterns: Vec<String>,
    pub test_types: Vec<QATestType>,
    pub options: TestGenerationOptions,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct TestGenerationOptions {
    pub include_edge_cases: bool,
    pub generate_mocks: bool,
    pub coverage_target: Option<f32>,
    pub test_framework: Option<String>,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct TestGenerationResponse {
    pub generated_files: Vec<GeneratedTestFile>,
    pub statistics: TestGenerationStats,
    pub warnings: Vec<String>,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct GeneratedTestFile {
    pub file_path: String,
    pub content: String,
    pub test_count: u32,
    pub coverage_estimate: f32,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct TestGenerationStats {
    pub total_files: u32,
    pub total_tests: u32,
    pub estimated_coverage: f32,
    pub generation_time_ms: u64,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct ConfigValidationResult {
    pub valid: bool,
    pub errors: Vec<String>,
    pub warnings: Vec<String>,
    pub suggestions: Vec<String>,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct QAStats {
    pub total_sessions: u64,
    pub total_tests_run: u64,
    pub average_coverage: f32,
    pub success_rate: f32,
    pub most_common_issues: Vec<IssueStats>,
    pub performance_trends: Vec<PerformanceTrend>,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct IssueStats {
    pub category: IssueCategory,
    pub count: u64,
    pub percentage: f32,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct PerformanceTrend {
    pub date: chrono::DateTime<chrono::Utc>,
    pub average_duration: f32,
    pub success_rate: f32,
}

#[derive(Debug, Clone)]
pub enum ExportFormat {
    Json,
    Xml,
    Html,
    Pdf,
}